### PR TITLE
914 make the lct secir model a derived class of flowmodel

### LINKED ISSUE #914: Make the LCT SECIR model a derived class of FlowModel
### Motivation / Current Behaviour
Currently the LCT model is not derived from a class. The functionality is similar to the FlowModel.
Also the files memilio/compartments/simulation.h for the CompartmentalModel and models/lct_secir/simulation.h for the LCT model do exactly the same. 
Both types of model should be combined.

### Enhancement description

It should be no problem to make the LCT SECIR model a derived class of FlowModel to have a consistent project structure.

### Additional context
For example: m_initial_values is used exactly like mio::CompartmentalModel::populations. If you want to, you could change its type to mio::Population, make the member public and rename it, so the models have more similar interface.

 _Originally posted by @reneSchm in https://github.com/SciCompMod/memilio/pull/941#discussion_r1502307064_

### Checklist

- [X] Attached labels, especially loc:: or model:: labels.
- [X] Linked to project

### PR DESCRIPTION
**Pull request by lenaploetzke**
# Changes and Information

Please **briefly list the changes** (main added features, changed items, or corrected bugs) made:

-  LCT model is now a derived class of CompartmentalModel
- Simulation function of CompartmentalModel is now used instead of separate function
- LctState works with size_t instead of int

If need be, add additional information and what the reviewer should look out for in particular:

-
-

## Merge Request - Guideline Checklist

Please check our [git workflow](https://github.com/SciCompMod/memilio/wiki/git-workflow). Use the **draft** feature if the Pull Request is not yet ready to review.

### Checks by code author

- [x] Every addressed issue is linked (use the "Closes #ISSUE" keyword below)
- [x] New code adheres to [coding guidelines](https://github.com/SciCompMod/memilio/wiki/Coding-guidelines)
- [x] No large data files have been added (files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] Tests are added for new functionality and a local test run was successful (with and without OpenMP)
- [x] Appropriate **documentation** for new functionality has been added (Doxygen in the code and Markdown files if necessary)
- [x] Proper attention to licenses, especially no new third-party software with conflicting license has been added
- [ ] (For ABM development) Checked [benchmark results](https://github.com/SciCompMod/memilio/wiki/Agent-Based-Model-Development) and ran and posted a local test above from before and after development to ensure performance is monitored.

### Checks by code reviewer(s)

- [ ] Corresponding issue(s) is/are linked and addressed
- [ ] Code is clean of development artifacts (no deactivated or commented code lines, no debugging printouts, etc.)
- [ ] Appropriate **unit tests** have been added, CI passes, code coverage and performance is acceptable (did not decrease)
- [ ] No large data files added in the whole history of commits(files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [ ] On merge, add 2-5 lines with the changes (main added features, changed items, or corrected bugs) to the merge-commit-message. This can be taken from the **briefly-list-the-changes** above (best case) or the separate commit messages (worst case).

Closes #914.

### CODE REVIEW

File: cpp/models/lct_secir/model.h [hunk 1]
```diff
@@ -186,119 +176,95 @@ class Model
...
-            return populations;
+            TimeSeries<ScalarType> compartments((Eigen::Index)InfectionState::Count);
+            Eigen::VectorXd wrong_size = Eigen::VectorXd::Constant(Eigen::Index(InfectionState::Count), -1);
+            compartments.add_time_point(-1, wrong_size);
+            return compartments;
         }
-        TimeSeries<ScalarType> populations((int)InfectionState::Count);
-        Eigen::VectorXd dummy((int)InfectionState::Count);
+        TimeSeries<ScalarType> compartments((Eigen::Index)InfectionState::Count);
+        Eigen::VectorXd dummy((Eigen::Index)InfectionState::Count);
```
Comment (reneSchm) [inline cpp/models/lct_secir/model.h:229]: How about naming this one ~Steve~ compartments, and the TimeSeries compartment_result(s) or compartment_ts, ...? Or without renaming the TS, how about (compartment(s)_)time_point, entry, ...?

File: cpp/models/lct_secir/model.h [hunk 2]
```diff
@@ -186,119 +176,95 @@ class Model
...
         for (Eigen::Index i = 0; i < result.get_num_time_points(); ++i) {
             // Use segment of vector of the result with subcompartments of InfectionState with index j and sum up values of subcompartments.
-            dummy[(int)InfectionState::Susceptible] = result[i][0];
-            dummy[(int)InfectionState::Exposed] =
+            dummy[Eigen::Index(InfectionState::Susceptible)] = result[i][0];
```
Comment (reneSchm) [inline cpp/models/lct_secir/model.h:232]: Here (and in the following accesses to ~Steve~ dummy) you use the ctor again, can you use casts instead?
Though it should be fine to use neither, the access operator should cause an implicit cast `size_t -> Index` (but it never hurts to make casts explicitly either).

**Review by reneSchm — CHANGES_REQUESTED** [review]
There are two exceedingly important requests regarding the poor dummy variables. Otherwise, this needs only minor changes, nice work.

File: cpp/models/lct_secir/model.h (hunk 2 above)
Comment (lenaploetzke) [reply cpp/models/lct_secir/model.h:232]: Yes you are absolutely right!

File: cpp/tests/test_lct_secir.cpp [hunk 3]
```diff
@@ -156,53 +167,69 @@ TEST(TestLCTSecir, compareWithOdeSecir)
...
+                    result_ode[i][Eigen::Index(mio::osecir::InfectionState::InfectedSymptoms)], 1e-5);
+        ASSERT_NEAR(0, result_ode[i][Eigen::Index(mio::osecir::InfectionState::InfectedSymptomsConfirmed)], 1e-5);
+        ASSERT_NEAR(result_lct[i][Eigen::Index(Model::LctState::InfectionState::InfectedCritical)],
+                    result_ode[i][Eigen::Index(mio::osecir::InfectionState::InfectedCritical)], 1e-5);
+        ASSERT_NEAR(result_lct[i][Eigen::Index(Model::LctState::InfectionState::InfectedSevere)],
+                    result_ode[i][Eigen::Index(mio::osecir::InfectionState::InfectedSevere)], 1e-5);
+        ASSERT_NEAR(result_lct[i][Eigen::Index(Model::LctState::InfectionState::Recovered)],
+                    result_ode[i][Eigen::Index(mio::osecir::InfectionState::Recovered)], 1e-5);
+        ASSERT_NEAR(result_lct[i][Eigen::Index(Model::LctState::InfectionState::Dead)],
+                    result_ode[i][Eigen::Index(mio::osecir::InfectionState::Dead)], 1e-5);
```
[the remark this answers is not in the document]
Comment (lenaploetzke) [reply cpp/tests/test_lct_secir.cpp:187]: Thank you for this reminder!

**Pull request merged**